srv/src/main/java/my/bookshop/handlers/AdminServiceHandler.java

```diff
@@ -320,4 +345,104 @@ private BigDecimal defaultZero(BigDecimal decimal) {
 		return decimal == null ? BigDecimal.valueOf(0) : decimal;
 	}
 
+	// audit logging
+
+	/**
+	 * Writes a data access message to the audit log for the given order if the order number is read.
+	 *
+	 * @param order the accessed order
+	 */
+	private void auditAccess(Orders order) {
+		if (order.getOrderNo() != null) {
+			this.auditLog.logDataAccess(createAccess(order));
+		}
+	}
+
+	/**
+	 * Writes a data modification message to the auditlog if the order number has changed.
+	 *
+	 * @param order the modified order
+	 */
+	private void auditChanges(Orders order) {
+		// check if order number is changed
+		if (order.getOrderNo() != null) {
+			// prepare a select statement to read old order number
+			Select<Orders_> ordersSelect = Select.from(ORDERS).columns(Orders_::OrderNo)
+					.where(o -> o.ID().eq(order.getId()).and(o.IsActiveEntity().eq(true)));
+
+			// read old order number from DB
+			Optional<Orders> oldOrderOpt = this.db.run(ordersSelect).first(Orders.class);
+
+			if (oldOrderOpt.isPresent()) {
+				Orders oldOrders = oldOrderOpt.get();
+
+				// check if there was a data modification
+				if (!StringUtils.equals(order.getOrderNo(), oldOrders.getOrderNo())) {
+					DataModification dataModification = createDataModification(order, oldOrders);
+					this.auditLog.logDataModification(Arrays.asList(dataModification));
+				}
+			}
+		}
+	}
+
+	private Access createAccess(Orders orders) {
+		Access access = Access.create();
+		access.setDataObject(createDataObject(orders));
+		access.setAttributes(createAttributes(Orders.ORDER_NO));
+		return access;
+	}
+
+	private static DataModification createDataModification(Orders orders, Orders oldOrders) {
+		ChangedAttribute attribute = createChangedAttribute(Orders.ORDER_NO, orders.getOrderNo(),
+				oldOrders.getOrderNo());
+
+		DataModification dataModification = DataModification.create();
+		dataModification.setDataObject(createDataObject(orders));
+		dataModification.setDataSubject(createDataSubject(oldOrders));
```

Comment:
Why do you use `oldOrders` and not `orders` like in the line before?

Reply:
Right, it's wrong will be fixed.